core/src/main/scala/kafka/raft/KafkaMetadataLog.scala

```diff
@@ -147,18 +221,102 @@ class KafkaMetadataLog(
   }
 
   override def createSnapshot(snapshotId: raft.OffsetAndEpoch): RawSnapshotWriter = {
-    FileRawSnapshotWriter.create(log.dir.toPath, snapshotId)
+    // Do not let the state machine create snapshots older than the latest snapshot
+    latestSnapshotId().ifPresent { latest =>
+      if (latest.epoch > snapshotId.epoch || latest.offset > snapshotId.offset) {
+        // Since snapshots are less than the high-watermark absolute offset comparison is okay.
+        throw new IllegalArgumentException(
+          s"Attemting to create a snapshot ($snapshotId) that is not greater than the latest snapshot ($latest)"
+        )
+      }
+    }
+
+    FileRawSnapshotWriter.create(log.dir.toPath, snapshotId, Optional.of(this))
   }
 
   override def readSnapshot(snapshotId: raft.OffsetAndEpoch): Optional[RawSnapshotReader] = {
     try {
-      Optional.of(FileRawSnapshotReader.open(log.dir.toPath, snapshotId))
+      if (snapshotIds.contains(snapshotId)) {
+        Optional.of(FileRawSnapshotReader.open(log.dir.toPath, snapshotId))
+      } else {
+        Optional.empty()
+      }
     } catch {
-      case e: NoSuchFileException => Optional.empty()
+      case _: NoSuchFileException =>
+        Optional.empty()
+    }
+  }
+
+  override def latestSnapshotId(): Optional[raft.OffsetAndEpoch] = {
+    try {
+      Optional.of(snapshotIds.last)
+    } catch {
+      case _: NoSuchElementException =>
+        Optional.empty()
+    }
+  }
+
+  override def oldestSnapshotId(): Optional[raft.OffsetAndEpoch] = {
+    oldestSnapshotId
+  }
+
+  override def onSnapshotFrozen(snapshotId: raft.OffsetAndEpoch): Unit = {
+    snapshotIds.add(snapshotId)
+  }
+
+  override def updateLogStart(logStartSnapshotId: raft.OffsetAndEpoch): Boolean = {
```

Comment:
One thing I was considering is whether we even want to expose this at all. In other words, who should be responsible for cleaning up old snapshots? I am tempted to say that that should be the responsibility of the Log implementation, but perhaps there are reasons it should not be?

Reply:
Yeah. I think it depends on the threading model. Right now I am relying on the polling thread to trigger the deletes. Note that the delete still happen async in the scheduler passed to the `kafka.log.Log`.

Having said that, when we implement https://issues.apache.org/jira/browse/KAFKA-12155 I think will have to give the Raft Client event more control as to how the log start offset gets incremented. For example, that Jira proposes looking at the `nextOffset` (acked offset) of the state machines and the `fetch offset` of the follower to increase the log start offset to minimize state machine invalidation.